src/engine/QueryPlanner.cpp

```diff
@@ -1319,6 +1334,7 @@ QueryPlanner::runDynamicProgrammingOnConnectedComponent(
                << std::endl;
     dpTab.emplace_back(vector<SubtreePlan>());
     for (size_t i = 1; i * 2 <= k; ++i) {
+      cancellationHandle_->throwIfCancelled("QueryPlanner producing plans");
```

Comment:
I think you also need to integrate checks into the `merge` method called for each pair of rows below (called in l. 1338) as in the exponential blowup case, there might be very long rows and a single one of those calls might be expensive.
That function is also just a nested loop where each inner iteration is expensive enough to check it.